proposals/0000-local-import.rst

```diff
@@ -0,0 +1,225 @@
+.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/0>`_.
+            **After creating the pull request, edit this file again, update the
+            number in the link, and delete this bold sentence.**
+.. sectnum::
+.. contents::
+
+This proposal allows the use of ``import`` statements in any ``let``, ``where`` or ``do`` block. The effect of those imports is limited to the corresponding block and (for ``let`` and ``where`` blocks) to the defined expression. Moreover, it defines a shortcut syntax, ``Qualifier.{ … }``, to allow scoping a local import over a single expression.
+
+For example, the following snippets are considered valid:
+
+::
+
+  module MyModule where
+
+  import Data.Set (Set)
+  import qualified Data.Set as Set
+  
+  foo :: (Set Int, Set String)
+  foo = (s1, fromList ["foo", "bar", "baz"])
+    where
+      import Set
+      s1 = fromList [1, 2, 3]
+
+::
+
+  module Main where
+  import System.Environment as Env
+   
+  main :: IO ()
+  main = do
+    args <- Env.getArgs
+    case args of
+      x:y:_ -> …
+      _ -> do
+        import qualified System.Exit as Exit
+        Exit.{ exitWith $ ExitFailure (-1) }
+
+Motivation
+------------
+This proposal is strongly inspired by two OCaml constructs, introduced in 2011 with version 3.12: ``let open Module in …`` and ``Module.( … )``. In less than 10 years, this feature has gained widespread adoption throughout the ecosystem, bringing significant benefits in terms of readability and maintainability.
+
+The proposed changes introduce Haskell equivalents in order to:
+
+- Make import lists smaller and easier to maintain.
+- Make it easier to understand why a module is imported, and where its symbols are used.
+- Make it easier to refactor a definition, along with its imports.
+- Make it more practical to use qualified imports, especially with operators.
+
+Import lists in Haskell can quickly become unwieldly. The recent proliferation of GHC proposals intending to simplify module imports is good evidence that this feeling is widely shared: 1_, 2_, 3_.
+
+Oftentimes, a module is only imported for a single symbol or two. The purpose of such imports is not obvious at first glance and can significantly clutter the import list. Moving those imports closer to the use-site would solve both of these problems by making the intent clearer and removing them from the toplevel list. Moreover, definitions would be easier to refactor, as they would be more self-contained ; for example, it would be easy to remove a definition along with its specific imports.
```

Comment:
```suggestion
Oftentimes, a module is only imported for a single symbol or two. The purpose of such imports is not obvious at first glance and can significantly clutter the import list. Moving those imports closer to the use-site would solve both of these problems by making the intent clearer and removing them from the toplevel list. Moreover, definitions would be easier to refactor, as they would be more self-contained; for example, it would be easy to remove a definition along with its specific imports.
```